File: src/main/java/moveZeroes/MoveZeroes.java
/*
 * @Source: https://leetcode-cn.com/problems/move-zeroes/
 * @Author: Tian Zhuo
 * @Date: 2019-01-10 21:05:42
 */
package moveZeroes;

public class MoveZeroes {
    public void moveZeroes(int[] nums) {
        int i = 0; // non-zero index
        int j = 0; // zero index
        while (i < nums.length) {
            if (nums[i] != 0) {
                i++;
            } else {
                if (j == 0) {
                    j = i;
                }
                while (nums[i] == 0) {
                    i++;
                    if (i == nums.length) {
                        return;
                    }
                }
                swap(nums, i, j);
                j++;
            }
        }
    }

    private void swap(int[] nums, int i, int j) {
        int t = nums[i];
        nums[i] = nums[j];
        nums[j] = t;
    }

    public void moveZeroes1(int[] nums) {
        int aa = 0;
        for (int i = 0; i < nums.length; i++) {
            if (nums[i] == 0) {
                aa++;
            } else {
                nums[i - aa] = nums[i];
            }

        }
        for (int i = nums.length - aa; i < nums.length; i++) {
            nums[i] = 0;
        }
    }

    public void moveZeroes2(int[] nums) {
        int i = 0;
        int start = 0;
        while (i < nums.length) {
            if (nums[i] != 0) {
                nums[start] = nums[i];
                if (start != i) {
                    nums[i] = 0;
                }
                start++;
            }
            i++;
        }
    }
}

File: algorithms/js/moveZeroes/moveZeroes.js
/*
 * @Source: https://leetcode-cn.com/problems/move-zeroes/
 * @Author: Tian Zhuo
 * @Date: 2019-01-10 23:26:56
 */
/**
 * @param {number[]} nums
 * @return {void} Do not return anything, modify nums in-place instead.
 */
var moveZeroes = function (nums) {
    let i = 0;
    let start = 0;
    while (i < nums.length) {
        if (nums[i] != 0) {
            nums[start] = nums[i];
            if (start != i) {
                nums[i] = 0;
            }
            start++;
        }
        i++;
    }
};

var moveZeroes2 = function (nums) {
    let zero = 0;
    let count = 0
    for (let i = 0; i < nums.length; i++) {
        if (nums[i] === zero) {
            nums.splice(i, 1);
            i--;
            count++
        }
    }
    for (let j = 0; j < count; j++) {
        nums.push(zero);
    }
};